File: solutions/python/diamond/diamond.py
"""Generate a Diamond pattern."""

import string


def rows(letter: str) -> list[str]:
    """Return a diamond pattern for a letter."""
    count = string.ascii_uppercase.index(letter)
    out = []
    for offset in range(0, count + 1):
        char = string.ascii_uppercase[offset]
        # Populate a space-filled line.
        line = [" "] * (2 * count + 1)
        # Insert the char, offset from the center.
        line[count - offset] = char
        line[offset - count - 1] = char
        out.append("".join(line))

    # Mirror the top half for the bottom.
    return out + list(reversed(out[:-1]))

File: solutions/go/diamond/diamond.go
package diamond

import (
	"fmt"
	"strings"
)

// Gen returns a diamond pattern string.
func Gen(char byte) (string, error) {
	if char < 'A' || char > 'Z' {
		return "", fmt.Errorf("invalid char %b not in range [A, Z]", char)
	}

	first := int('A')
	last := int(char)
	length := last - first
	grid := make([]string, length*2+1)

	// Top row
	grid[0] = strings.Repeat(" ", length) + "A" + strings.Repeat(" ", length)
	// B through char.
	for i := 1; i <= length; i++ {
		edge := strings.Repeat(" ", length-i)
		center := strings.Repeat(" ", i*2-1)
		grid[i] = edge + string(byte(first+i)) + center + string(byte(first+i)) + edge
	}
	// Mirror the bottom half.
	for i := length; i >= 0; i-- {
		grid[2*length-i] = grid[i]
	}

	return strings.Join(grid, "\n"), nil
}
